### ISSUE TITLE
rook-ceph upgrade from 1.5.9 to 1.6.0: CustomResourceDefinition "cephblockpools.ceph.rook.io" is invalid: spec.preserveUnknownFields

### ISSUE THREAD

**Issue by ibotty** [rook/rook#7659]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
```
> kubectl apply -f crds.yaml
The CustomResourceDefinition "cephblockpools.ceph.rook.io" is invalid: spec.preserveUnknownFields: Invalid value: true: must be false in order to use defaults in the schema
```

**How to reproduce it (minimal and precise):**

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod): 1.5.9
* Storage backend version (e.g. for ceph do `ceph -v`):
* Kubernetes version (use `kubectl version`): Server Version: 4.7.0-0.okd-2021-04-11-124433, Kubernetes Version: v1.20.0-1056+c8905da1d144e4-dirty
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Openshift
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

**Comment by subhamkrai** [comment]
@ibotty please follow this [doc](https://rook.github.io/docs/rook/v1.6/ceph-upgrade.html) to upgrade. Thanks

**Comment by ibotty** [comment]
I did.  `kubectl apply -f common.yaml -f crds.yaml` is the step that fails with

```
The CustomResourceDefinition "cephblockpools.ceph.rook.io" is invalid: spec.preserveUnknownFields: Invalid value: true: must be false in order to use defaults in the schema
```

This installation has been steadily upgraded from Rook v1.3 afair so it might be some leftover.  Should I `replace -f`?

**Comment by ibotty** [comment]
I had to remove `spec.preserveUnknownFields` from most `*.ceph.rook.io` CRDs manually.  I could apply `crds.yaml` after doing that.

**Comment by travisn** [comment]
The `spec.preserveUnknownFields` seems to be an artifact leftover from converting from the v1beta1 to v1 CRDs. In v1  the tag to preserve unknown fields is `x-kubernetes-preserve-unknown-fields: true`. We made this conversion in v1.5, so I'm unclear why you are hitting this with v1.6 and not previously. Good to hear you could work around it by removing that from your CRDs, but that certainly is tedious so I'd like to find a better way.

**Comment by mankuen** [comment]
I encountered the same situation, which was also solved by remove spec.preserveUnknownFields from *.ceph.rook.io CRDs manually. It seems to be left over.

**Comment by psavva** [comment]
I have faced the same issue

**Comment by BlaineEXE** [comment]
My findings:
- using `kubectl replace` works but will not create new resources
- using `kubectl apply` creates new resources but fails with this bug
- using `kubectl apply --force` deletes and then re-creates resources
  - this could be a good solution, or it could cause accidental deletion of custom resources **NEED TO INVESTIGATE THIS**

**Comment by SerialVelocity** [comment]
@BlaineEXE @leseb How can I apply the workaround when using Helm? Since the CRDs are owned by Helm, I can't just apply the CRD yaml from your examples folder.

**Comment by oaudry** [comment]
@SerialVelocity kubectl edit crd cephobjectzones.ceph.rook.io  and set the field spec.preserveUnknownFields to false and restart helm upgrade should do the trick

**Comment by loranmutafov** [comment]
For anyone coming into this with a non-Helm setup, here's how I fixed it. It's basically what @BlaineEXE said, but `kubect replace`ing just the `crds.yaml`, as `common.yaml` defines the namespace and I don't wanna risk deleting this.

**First I did a** 
```sh
kubectl diff -f common.yaml -f crds.yaml
```

which gave me the same error as above

**I decided to see how far it'll go with an apply**
```sh
kubectl apply -f common.yaml -f crds.yaml
```

and it applied everything from `common.yaml`, and failed on `crds.yaml`, with the same error message as in the thread.

**So next I ran:**

```sh
kubectl replace -f crds.yaml
kubectl diff -f common.yaml -f crds.yaml
```

**The diff gave me no changes, so I decided to run an apply again**
```sh
kubectl apply -f common.yaml -f crds.yaml
```

which complained about some annotations, but it said it'll patch them automatically.

**To ensure it had indeed patched them, I ran another apply**
```sh
kubectl apply -f common.yaml -f crds.yaml
```
and this one indeed returned no errors


And yes, the cluster is alright
